File: src/main/python/com/hazza/leetcode_by_python/SegmentOne/No44_Wildcard_Matching.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/5/14 20:12
# https://leetcode.com/problems/wildcard-matching/


class Solution:
    # solution 1
    def isMatch1(self, s: str, p: str) -> bool:
        lenS, lenP = len(s), len(p)
        si, pi, start, match = 0, 0, -1, 0

        while si < lenS:
            if pi < lenP and (s[si] == p[pi] or p[pi] == '?'):
                si += 1
                pi += 1
            elif pi < lenP and p[pi] == '*':
                start = pi
                match = si
                pi += 1
            elif start != -1:
                pi = start + 1
                match += 1
                si = match
            else:
                return False

        return pi == lenP or p.count('*', pi) == lenP - pi

    # solution 2
    def isMatch2(self, s: str, p: str) -> bool:
        lenP = len(p)
        dp = [False] * (lenP + 1)

        dp[0] = True
        for i in range(1, lenP + 1):
            dp[i] = dp[i - 1] and p[i - 1] == '*'

        for c in s:
            prev = dp[0]
            for j in range(1, lenP + 1):
                temp = dp[j]
                if c == p[j - 1] or p[j - 1] == '?':
                    dp[j] = prev
                elif p[j - 1] == '*':
                    dp[j] = dp[j] or dp[j - 1]
                else:
                    dp[j] = False
                prev = temp
            dp[0] = False

        return dp[lenP]


if __name__ == '__main__':
    solution = Solution()

    assert solution.isMatch1('aa', 'a') is False
    assert solution.isMatch1('aa', '*') is True
    assert solution.isMatch1('cb', '?a') is False
    assert solution.isMatch1('adceb', '*a*b') is True
    assert solution.isMatch1('acdcb', 'a*c?b') is False


    assert solution.isMatch2('aa', 'a') is False
    assert solution.isMatch2('aa', '*') is True
    assert solution.isMatch2('cb', '?a') is False
    assert solution.isMatch2('adceb', '*a*b') is True
    assert solution.isMatch2('acdcb', 'a*c?b') is False

File: src/main/java/com/hazza/leetcode/SegmentOne/No44WildcardMatching.java
package com.hazza.leetcode.SegmentOne;

/**
 * https://leetcode.com/problems/wildcard-matching/discuss/
 */
public class No44WildcardMatching {
    // solution 1
    public boolean isMatch(String s, String p) {
        if (p.length() == 0) return s.length() == 0;
        int si = 0, pi = 0, start = -1, match = 0;
        int lenS = s.length(), lenP = p.length();
        while (si < lenS) {
            if (pi < lenP && (s.charAt(si) == p.charAt(pi) || p.charAt(pi) == '?')) {
                ++si;
                ++pi;
            } else if (pi < lenP && p.charAt(pi) == '*') {
                start = pi;
                match = si;
                ++pi;
            } else if (start != -1){
                pi = start + 1;
                ++match;
                si = match;
            } else return false;
        }

        while (pi < lenP && p.charAt(pi) == '*') ++pi;
        return pi == lenP;
    }

    // solution 2.1
    public boolean isMatch2(String s, String p) {
        if (p.length() == 0) return s.length() == 0;
        int lenS = s.length(), lenP = p.length();
        boolean[][] m = new boolean[lenS + 1][lenP + 1];

        m[0][0] = true;
        for (int i = 1; i <= lenP; i++)
            m[0][i] = m[0][i - 1] && p.charAt(i - 1) == '*';

        for (int i = 1; i <= lenS; i++) {
            for (int j = 1; j <= lenP; j++) {
                if (s.charAt(i - 1) == p.charAt(j - 1) || p.charAt(j - 1) == '?')
                    m[i][j] = m[i - 1][j - 1];
                else if (p.charAt(j - 1) == '*')
                    m[i][j] = m[i - 1][j] || m[i][j - 1];
            }
        }
        return m[lenS][lenP];
    }

    // solution 2.2
    public boolean isMatch3(String s, String p) {
        if (p.length() == 0) return s.length() == 0;
        int lenS = s.length(), lenP = p.length();
        boolean[] m = new boolean[lenS + 1];
        m[0] = true;
        for (int j = 1; j <= lenP; j++) {
            boolean pre = m[0];
            m[0] = m[0] && p.charAt(j - 1) == '*';
            for (int i = 1; i <= lenS; i++) {
                boolean temp = m[i];
                if (s.charAt(i - 1) == p.charAt(j - 1) || p.charAt(j - 1) == '?')
                    m[i] = pre;
                else if (p.charAt(j - 1) == '*')
                    m[i] = (m[i] || m[i - 1]);
                else m[i] = false;
                pre = temp;
            }
        }
        return m[lenS];
    }
}
